text/3607-enum-discriminant.md

```diff
@@ -0,0 +1,413 @@
+*Note: this is different from `mem::discriminant`.  For example,*
+```rust
+#![allow(enum_intrinsics_non_enums)]
+enum MyEnum { A, B }
+let a = Box::new(MyEnum::A);
+let b = Box::new(MyEnum::B);
+assert_eq!(std::mem::discriminant(&a), std::mem::discriminant(&b));
+assert_ne!(a.enum#discriminant, b.enum#discriminant);
+```
+<!-- https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=76293c2f83cb7719c22c15bcebeaeb13 -->
+
+For this, a generic parameter is never considered to be an `enum`,
+although a generic enum where some of the generic parameters to the
+enum constructor are not yet known is fine.
+
+It's an error if, despite deref'ing, the LHS is still not an `enum`.
+
+If the enum has `repr(uN)` or `repr(iM)`, the `.enum#discriminant` expression
+returns a value of type `uN` or `iM` respectively.
+
+If the enum does not specify an integer `repr`, then it returns `isize`.
+
+*Note: `isize` is rarely the desired type for discriminants, and indeed custom
+discriminants on types with fields are disallowed without explicit `repr` types.
+Returning `isize` is fine here, though, thanks to privacy because the code
+inside the module can be updated should it change to specify a specific type.*
+
+## Semantics
+
+When the LHS of a discriminant expression is a *place*, that place is read but not consumed.
+
+*Note: this can be thought of as if it read a field of `Copy` type from the LHS.*
+
+This lowers to [`Rvalue::Discriminant`][MIR discr] in MIR.
+
+[MIR discr]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.Rvalue.html#variant.Discriminant
+
+As this expression is an *r-value*, not a *place*, `&foo.enum#discriminant` returns
+a reference to a temporary, aka is the same as `&{foo.enum#discriminant}`.
+It does *not* return a reference to the memory in which the discriminant is
+stored -- not even for types that do store the discriminant directly.
+
+This expression is allowed in `const` contexts, but is not promotable.
+
+*Note: the behaviour of this expression is independent of whether the type gets
+layout-optimized.  For example, the following holds even if `x` is `2_i8` in memory.*
+```rust
+enum MyOption<T> { MyNone, MySome(T) }
+let x = MyOption::<std::cmp::Ordering>::MyNone;
+assert_eq!(x.enum#discriminant, 0_isize);
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This isn't strictly necessary, we could continue to get along just fine without it.
+
+- For the FFI cases the layout guarantees mean it's already possible to write a
+  sound and reliable function that reads the discriminant.
+- For cases without `repr(int)`, custom discriminants aren't even allowed,
+  so those discriminants much not be all that important.
+- It's always possible to write a `match` in safe code that optimizes away
+  and produces exactly the same thing that this new expression would.
+- A pseudo-field with `#` in the name looks kinda weird.
+- There might be a nicer way to do this in the future.
+
+# Rationale and alternatives
```

Comment:
_Implicitly_-assigned enum discriminants are a rare case in Rust where the lexical ordering of definitions has an implicit, observable and (potentially) SemVer-relevant effect. It's arguably a footgun. This RFC will increase the range of situations in which implicitly-assigned discriminants are observable.

In the interest of starting to nudge folks away from this footgun, perhaps this proposed accessor should be restricted to enums with strictly explicit discriminants?

This restriction might be sensible to adopt as a first step. It will be easier to relax this restriction later, than it will be to add it.

Reply:
> In the interest of starting to nudge folks away from this footgun, perhaps this proposed accessor should be restricted to enums with strictly explicit discriminants?

It's a no-go until Rust allow assigning an explicit discriminant to variants of a non-primitive representation enum (i.e. plain `#[repr(Rust)]` without any `#[repr(uNN)]`). 

Forcing `#[repr(uNN)]` means any user of this RFC must forego layout optimization, but this is not an intrinsic restriction since `std::intrinsic::discriminant_value()` does work for every kind of enums.

Reply:
> It's a no-go until Rust allow assigning an explicit discriminant to variants of a non-primitive representation enum (i.e. plain `#[repr(Rust)]` without any `#[repr(uNN)]`).

That's totally on the table! I only did not implement that because there was no way to reflect on such discriminants. With this RFC, we almost certainly should remove that restriction.

Reply:
Good to hear, if there is a formal proposal or issue somewhere it could be added as a prerequisite :smiley: 

But given the motivation, do we expect enum authors in e.g. Edition 2027 must explicitly assign a variant in order to use `#[derive(PartialOrd)]`? (or, considering 3rd-party proc macros, `#[derive_ex(PartialOrd)]` / `#[derive_where(PartialOrd)]`)

Reply:
> if there is a formal proposal or issue somewhere it could be added as a prerequisite 

Yeah, it's RFC2363! That RFC proposes that explicit discriminants be applicable to *all* enums, even default repr ones. We just didn't implement that yet, because it wasn't possible to soundly reflect on those discriminants.

> But given the motivation, do we expect enum authors in e.g. Edition 2027 must explicitly assign a variant in order to use `#[derive(PartialOrd)]`?

Perhaps! Requiring that would also solve the common footgun that enum variants cannot be automatically reordered and alphabetized without semantic implications. But I could also see it continue being the case that we put off such an edition gated reqiremen on our first party derives indefinitely. A decision here would be outside the scope of this RFC, but should perhaps be mentioned as a future possibility.

Reply:
> Yeah, it's RFC2363! That RFC proposes that explicit discriminants be applicable to all enums, even default repr ones. We just didn't implement that yet, because it wasn't possible to soundly reflect on those discriminants.

seems not in [2363-proper](https://github.com/rust-lang/rfcs/pull/2363#issuecomment-375075559), it is listed as an "[unresolved question](https://github.com/rust-lang/rfcs/blob/master/text/2363-arbitrary-enum-discriminant.md#should-this-apply-only-to-enumerations-with-an-explicit-representation)".

Reply:
Yes, but the text of the RFC itself doesn't include that restriction, we already have an accepted design doc (and relevant discussions) for this point in the design space. 

Given that this question was only raised for lack of a reflection mechanism at the time (which this RFC will resolve), @scottmcm, perhaps let's formally resolve that question in this RFC? A sentence in the summary should suffice: "The implementation of RFC2363 shall also be extended to allow declaring the discriminant of default-repr enums."

Reply:
The RFC restricts visibility to the current module. So it still allows local reasoning. E.g. anything that tries to sort enum variants can look at the module and check if this syntax is used. It's worse than only having to look at the type but at least not spooky action at the distance.